cohort-engine/src/main/java/com/ibm/cohort/engine/measure/ResourceResolutionProvider.java

```diff
@@ -0,0 +1,298 @@
+ * may not be suitable for environments where the knowledge associated with
+ * these artifacts might change without revving the unique identifiers.
+ */
+public abstract class ResourceResolutionProvider
+		implements LibraryResolutionProvider<Library>, MeasureResolutionProvider<Measure> {
+	private Map<String, Map<String, MetadataResource>> resourcesByIdByResourceType = new HashMap<>();
+	private Map<String, Map<String, SortedMap<SemanticVersion, MetadataResource>>> resourcesByNameByResourceType = new HashMap<>();
+	private Map<String, Map<String, SortedMap<SemanticVersion, MetadataResource>>> resourcesByUrlByResourceType = new HashMap<>();
+	
+	/**
+	 * Process a FHIR resource. This parses the resource, resolves the unique
+	 * identifiers, and adds the resource to the underlying query caches.
+	 * 
+	 * @param resourceName Artifact name (e.g. filename, zip entry name, etc.) of
+	 *                     the resource that is being loaded. In circumstances where
+	 *                     no Resource.id is provided, this will serve as a
+	 *                     stand-in.
+	 * @param resourceData Stream containing the input data to be processed.
+	 * @param parser       HAPI FHIR resource parser.
+	 * @throws IOException when resource data cannot be read from the stream.
+	 */
+	protected void processResource(String resourceName, InputStream resourceData, IParser parser) throws IOException {
+		IBaseResource resource = parser.parseResource(resourceData);
+		processResource( resourceName, resource);
+	}
+	
+	/**
+	 * Process a FHIR resource. This parses the resource, resolves the unique
+	 * identifiers, and adds the resource to the underlying query caches.
+	 * 
+	 * @param resourceName Artifact name (e.g. filename, zip entry name, etc.) of
+	 *                     the resource that is being loaded. In circumstances where
+	 *                     no Resource.id is provided, this will serve as a
+	 *                     stand-in.
+	 * @param resource	   deserialized FHIR resource object
+	 * @throws IOException when resource data cannot be read from the stream.
+	 */
+	protected void processResource(String resourceName, IBaseResource resource) throws IOException {
+
+		if (resource instanceof MetadataResource) {
+			MetadataResource metadataResource = (MetadataResource) resource;
+
+			if (metadataResource.getId() == null) {
+				metadataResource.setId(FilenameUtils.getBaseName(resourceName));
+			}
+
+			addToIndexes(metadataResource);
+		} else if (resource instanceof Bundle) {
+			Bundle bundle = (Bundle) resource;
+
+			for (BundleEntryComponent be : bundle.getEntry()) {
+				if (be.getResource() instanceof MetadataResource) {
+					MetadataResource metadataResource = (MetadataResource) be.getResource();
+					if (metadataResource.getId() == null) {
+						throw new IllegalArgumentException("Bundle resources must contain an ID value");
+					}
+
+					addToIndexes(metadataResource);
+				}
+			}
+		}
```

Comment:
does it makes sense to:
```java
throw UnsupportedSomethingExeception();
```
here?

Reply:
Yes, good catch. We should throw an error if the resource type isn't handled.